### PR TITLE
fix/jwtSync

### LINKED ISSUE #68: [Bug] Expired token unable to refresh
```
2021-07-29 19:44:31.615  INFO [my-bot,1efb3e9c675e3506,1efb3e9c675e3506] 1 --- [nio-8080-exec-7] c.s.github.v3.clients.GitHubClient       : Github token for installation 1300 is either expired or null. Trying to get a new one.
--
2021-07-29 19:44:31.615  INFO [my-bot,1efb3e9c675e3506,1efb3e9c675e3506] 1 --- [nio-8080-exec-7] c.s.github.v3.clients.GitHubClient       : Got JWT Token. Now getting Github access_token for installation 1234
2021-07-29 19:44:31.619 ERROR [my-bot,a1558e2c87f905cd,a1558e2c87f905cd] 1 --- [nio-8080-exec-5] c.comp.path.WebhookErrorHandler  : error processing request: (Could not generate access token for github app)
java.lang.RuntimeException: Could not generate access token for github app
...
Caused by: java.lang.Exception: Got non-2xx status 401 when getting an access token from GitHub: Unauthorized
    at com.spotify.github.v3.clients.GitHubClient.generateInstallationToken(GitHubClient.java:607) ~[github-client-0.0.25.jar:0.0.25]
    at com.spotify.github.v3.clients.GitHubClient.getInstallationToken(GitHubClient.java:581) ~[github-client-0.0.25.jar:0.0.25]
    at com.spotify.github.v3.clients.GitHubClient.getAuthorizationHeader(GitHubClient.java:560) ~[github-client-0.0.25.jar:0.0.25]
```
on restart of the application; everything works fine.

### PR DESCRIPTION
**Pull request by bguedel**
fixes #68

per example in related issue; time sync issue can occur when setting issued at to be the current time. This sets the issued at to be 1 minute earlier to avoid this.

Tested locally for 2 days with no issues.

### CODE REVIEW

**Comment by bguedel** [comment]
@henriquetruta thoughts?

**Comment by bguedel** [comment]
@neonal Thoughts?

**Comment by bguedel** [comment]
@henriquetruta thoughts?

i see you merged a couple of other PRs that were opened after mine. Wondering when we can get this merged; This is a blocker for my team to use this project ;(

**Comment by bguedel** [comment]
@henriquetruta @neonal 

any thoughts?

**Comment by bguedel** [comment]
@neonal?? Looks like you merged a PR a few hours ago. Wondering if you can take a look at this PR??

Or can we get some guidance on if you guys actually accept PR from external?

**Comment by bguedel** [comment]
@barbatron Sorry for the delay;

Added the supplier. I defaulted the main supplier to be the same as it is today so that people that update dont see any change to their code;

Added the "60000" to the tests; but not sure what we expect to be tested here; Effectively i am testing to see if a jwt is created when using the default and using a custom supplier; If you have any other ideas on what can be tested i am open to suggestion.

Right now the output is just a string; sooooo

**Comment by shaendler** [comment]
@bguedel i also see this issue quite frequently in my app. using the 0.1.24 version.
Can you explain how as the library consumer i can control the isseuAt supplier? on my end, i am just creating a client

@barbatron can we urge this fix to merge?

**Comment by bguedel** [comment]
@shaendler good call;

https://github.com/spotify/github-java-client/blob/dbabfc3b4056797bf01b4d924d229f09bbdd2b49/src/main/java/com/spotify/github/v3/clients/GitHubClient.java#L622 <--- thats where this is actually being used.

which is essentially used by a request such as
https://github.com/spotify/github-java-client/blob/dbabfc3b4056797bf01b4d924d229f09bbdd2b49/src/main/java/com/spotify/github/v3/clients/GitHubClient.java#L361

which is used by the GithubClient --> to which you create in your app.

so this solution wont actually work; though it gives some building blocks to work.

Open to suggestion on how we can make it work. Dont love pushing this all the way up to the GithubClient factory methods.

Otherwise we go back to hardcoding 60000 (60s)???

**Comment by bguedel** [comment]
This is currently how we are using this (through those links)
```java
//Spring @Configuration file
/**
     * Spotify github client.
     *
     * @return
     */
    @Bean
    public GitHubClient githubClient() throws IOException {
        byte[] decodedBytes = Base64.getDecoder().decode(gitHubProperties.getApp().getPrivateKey());
        FileUtils.writeByteArrayToFile(new File(PK_PATH), decodedBytes);

        return GitHubClient.create(
                URI.create("https://pathTo.github/api/v3"),
                FileUtils.getFile(PK_PATH),
                gitHubProperties.getApp().getAppId(),
                gitHubProperties.getApp().getInstallationId());
    }
```

**Comment by bguedel** [comment]
ill make a pr right quick to change the default behavior to -60s and then still allow someone to override this;

i dont forsee this being an issue for current systems; this will solve the issue; but still give the chance to set this directly; if later someone wants to implement a way to use this feature deeper into the system.

File: src/main/java/com/spotify/github/v3/clients/JwtTokenIssuer.java [hunk 1]
```diff
@@ -30,37 +30,55 @@
...
 import java.util.Date;
+import java.util.function.Supplier;
 
 /** The helper Jwt token issuer. */
 public class JwtTokenIssuer {
 
   private static final SignatureAlgorithm SIGNATURE_ALGORITHM = SignatureAlgorithm.RS256;
-  private static final long TOKEN_TTL = 600000;
+  private static final long TOKEN_TTL = 600_000;
+  private static final long TOKEN_ISSUED = 60_000;
```
Comment (bguedel) [inline src/main/java/com/spotify/github/v3/clients/JwtTokenIssuer.java:40]: Fun Fact: Java totally allows _ in numbers for readability;

Comment (bguedel) [reply src/main/java/com/spotify/github/v3/clients/JwtTokenIssuer.java:40]: [linky](https://docs.oracle.com/javase/7/docs/technotes/guides/language/underscores-literals.html)

**Comment by bguedel** [comment]
built this locally again and put it in my app; getting the error again; this supplier doesnt seem to be working the way i had expected.

**Comment by wsguede** [comment]
Locally its also failing with the code i had before so thats weird; but this should be the fix. I have added duration testing as well and it does seem to work just fine.

This should be safe to accept.

**Comment by wsguede** [comment]
Looking at my old code that did work; the drift was only 6 seconds, but the jwt was only 1 minute; so i think the issue that i had during this fix was that the jwt was officially good for 11 minutes which i dont think should be correct. 

[example](https://github.com/gr2m/universal-github-app-jwt/blob/fae43d31a3abefd49deec849ee2dde1e96bcd65f/src/index.ts#L15) that i am pretty sure probot uses under the hood.

**Comment by bguedel** [comment]
so far this has worked locally; no more 401s

**Comment by bguedel** [comment]
@barbatron Thoughts?

**Comment by bguedel** [comment]
@barbatron Did you forget to submit your review again?

**Comment by bguedel** [comment]
@barbatron hope you are having a good december;

Before you cut out for some awesome end of year bash, could i get a comment or 4, or maybe event an approval?

dont know why my pr is not really getting built/codecov

**Comment by wsguede** [comment]
Happy New Year Spotify team!

Could someone take a look at this and give some feedback or accept the pr?

@barbatron @henriquetruta @neonal

**Comment by bguedel** [comment]
Hey Spotify team!

Just a friendly reminder; Looking to get this into master. Could I get some comments?
@barbatron @henriquetruta @neonal

**Comment by barbatron** [comment]
Hey there @bguedel , I'm no longer with Spotify so I'm afraid I can't be of help. Filtered out emails from the Spotify org due to the vast amount from other FOSS projects. I'm impressed with your tenacity, though! My team at Sp had lots to do and was getting even more on the plate, and without much buy-in to maintain these peripheral repos apart from dwindling spare time, I'm afraid it could be tricky to get attention. Best of luck though. Sorry I didn't get this ashore before I left back in October.

**Review by barbatron — APPROVED** [review]

**Comment by bguedel** [comment]
Hey @henriquetruta @neonal

Just wondering if you had any input on this change?

**Comment by bguedel** [comment]
Wondering if we can get some eyes on this?

**Comment by ebk45** [comment]
Hi @bguedel.

We apologise for taking so long to get eyes on this PR, we haven't been able to maintain this library to the standard we would have liked to but this will be changing in the new year. If you could please get this rebased and we'll be prioritising all open PRs ready for review.

Thanks!

Ellie

**Comment by ebk45** [comment]
Hello,

As this PR is now stale and we've not had any new activity on it, we've decided to close it so we can keep track of active/more recent changes. We will leave the branch untouched in case you would like to reopen and rebase.

Thanks!

Ellie

**Pull request closed without merging**